Fix dev/core#3682:  Add support for sitewide tokens such as "logo/header" in message templates, and mechanism to support others

twomice opened the pull request:
Overview
----------------------------------------
Per [dev/core#3682](https://lab.civicrm.org/dev/core/-/issues/3682) , we're looking for a way to support customizable tokens which could be used, for example, to insert an organization logo into message templates.

Taking inspiration (but a different approach) from @eileenmcnaughton 's #30441, this PR tackles this problem by:

* defining a new entity type "Site Token" (and sensible UI tools for managing those;
* providing one Site Token entity called "message_header" with an empty value; and 
* inserting that token into all relevant message templates.

Before
----------------------------------------
To insert a custom logo or other branding into Message Templates, the site admin must manually edit each such template, which then creates the administrative burden of diffing/editing those templates each time a CiviCRM upgrade makes changes elsewhere in the template.

After
----------------------------------------
A site admin can insert a custom logo or other branding into Message Templates simply by editing the value of the {site.message_header} Site Token. This leaves the Message Template itself unchanged, relieving the site admin of related administrative burden at upgrade time.

Technical Details
----------------------------------------
Generally, I'll be aiming for these changes:

- New entity type: Sitetoken
  - fields:
    - id
    - domain_id
    - name (token string, e.g. {site.[name]}
    - label (user-visible label in token UI)
    - body_html
    - body_text
    - is_active
    - is_reserved (cannot delete; cannot edit name)
    - created_id
    - modified_id
    - modified_date (REFERENCE: same as Mailing.modified_date)
  - validation (rather, enforcement via pre hook):
    - do not allow change to `name` field if `is_reserved = 1`
    - do not allow delete if `is_reserved = 1`
- New Sitetoken entity
  - name: 'message_header'
  - label: 'Message Header'
  - body_html: `'<!-- This is the {site.message_header} token HTML content. -->'`
  - body_text: `''`
  - is_reserved: 1
- New page/listing: Site Tokens
  - Purpose: List all existing Site Tokens, provide access to CRUD actions
  - Access: "CiviCRM: edit system workflow message templates"
  - mechanism: SearchKit
- New menu item: Administer > Communications > Site Tokens
  - Access: "CiviCRM: edit system workflow message templates"
  - Points to "Site Tokens" page/listing
- New form: Create / Edit Site Token
  - Access: "CiviCRM: edit system workflow message templates"
  - Mechanism: FormBuilder
- New Token group "Site Tokens":
  - prefix: site (e.g. {site.message_header})
  - Tokens in this group are as found in:  
    ```
      Civi\Api4\SiteToken::get(FALSE)
 	->addWhere('is_active', '=', 1');
    ```
- Changes to all (relevant?) message templates to:
  - use the `{site.message_header}` token
  - include a Smarty comment pointing to documentation on this new functionality (instead of the current ` <!-- You can add table row(s) here with logo or other header elements -->`)
- New documentation explaining this functionality
  - Probably at https://docs.civicrm.org/user/en/latest/email/message-templates/#modifying-system-workflow-message-templates
- New unit tests as appropriate.
    

Comments
----------------------------------------
Aiming to have this done during the "Civi Beach Week" sprint.

twomice committed:
Fix dev/core#3682 Support Site Tokens.

eileenmcnaughton commented:
@mattwire @kurund just pinging you to flag this PR as the current effort on this - which as you pointed out overlaps with https://lab.civicrm.org/extensions/prettyworkflowmessages

I *think* it's probably a good thing that the namespace is different - even though some of the functionality is the same - ie people already using the extension won't be affected by this unless they choose to migrate to using the new tokens

twomice commented on ang/afform/afformSiteToken.aff.html:
```diff
@@ -0,0 +1,13 @@
+  <af-entity data="{source: 'Site Token'}" type="SiteToken" name="SiteToken1" label="Site Token 1" actions="{create: true, update: true}" security="FBAC" url-autofill="1" />
+  <fieldset af-fieldset="SiteToken1" class="af-container">
+    <afblock-name-sitetoken></afblock-name-sitetoken>
+    <af-field name="name" defn="{required: true, input_attrs: {}}" />
```
@colemanw or @totten : This form is to edit/create a site token, and I'd like the Name field to be read-only if `is_reserved = 1`. Any thoughts on how to do that in formBuilder?

eileenmcnaughton commented on CRM/Upgrade/Incremental/schema/5.76.alpha1.SiteToken.entityType.php:
```diff
@@ -0,0 +1,155 @@
+      'title' => ts('Token Name'),
+      'sql_type' => 'varchar(64)',
+      'input_type' => 'Text',
+      'description' => ts('Token string, e.g. {site.[name]}'),
```
```suggestion
      'description' => ts('Token string, e.g. {site.[name]}'),
      'required' => TRUE,
```

eileenmcnaughton commented on CRM/Upgrade/Incremental/schema/5.76.alpha1.SiteToken.entityType.php:
```diff
@@ -0,0 +1,155 @@
+      'title' => ts('Token Label'),
+      'sql_type' => 'varchar(255)',
+      'input_type' => 'Text',
+      'description' => ts('User-visible label in token UI'),
```
```suggestion
      'description' => ts('User-visible label in token UI'),
      'required' => TRUE,
```

eileenmcnaughton commented on CRM/Upgrade/Incremental/schema/5.76.alpha1.SiteToken.entityType.php:
```diff
@@ -0,0 +1,155 @@
+      'title' => ts('Token Value (Text)'),
+      'sql_type' => 'text',
+      'input_type' => 'TextArea',
+      'description' => ts('Value of the token in text format.'),
```
I wonder if we can make it clear that this may not be needed as it will be inferred if not present

eileenmcnaughton commented on CRM/Upgrade/Incremental/schema/5.76.alpha1.SiteToken.entityType.php:
```diff
@@ -0,0 +1,155 @@
+      'sql_type' => 'int unsigned',
+      'input_type' => 'EntityRef',
+      'description' => ts('Contact responsible for creating this token'),
+      'add' => '4.3',
```
```suggestion
      'add' => '5.76',
```

eileenmcnaughton commented on CRM/Upgrade/Incremental/schema/5.76.alpha1.SiteToken.entityType.php:
```diff
@@ -0,0 +1,155 @@
+      'input_type' => 'Select Date',
+      'readonly' => TRUE,
+      'description' => ts('When the token was created or modified or deleted.'),
+      'add' => '4.7',
```
```suggestion
      'add' => '5.67',
```

eileenmcnaughton commented on CRM/Core/DAO/SiteToken.php:
```diff
@@ -0,0 +1,12 @@
+
+/**
+ * Placeholder class retained for legacy compatibility.
+ */
```
```suggestion
  * @property int $id
 * @property int $domain_id
 * @property string $name
 * @property string $body_html
 * @property string $body_text
 * @property bool $is_active
 * @property bool $is_reserved
 * @property int $created_id
 * @property int $modified_id
 * @property string $modified_date
 */
```

eileenmcnaughton commented on Civi/Core/Container.php:
```diff
@@ -389,6 +389,10 @@ public function createContainer() {
       'CRM_Core_DomainTokens',
       []
     ))->addTag('kernel.event_subscriber')->setPublic(TRUE);
+    $container->setDefinition('crm_site_tokens', new Definition(
```
It's not quite merged yet but when this is merged https://github.com/civicrm/civicrm-core/pull/30461 it won't be necessary to do this as it can be declared in the entityType

eileenmcnaughton commented on schema/Core/SiteToken.entityType.php:
```diff
@@ -0,0 +1,155 @@
+<?php
+
+return [
+  'name' => 'SiteToken',
```
oh - so all my comments above should have been / apply to this

eileenmcnaughton commented:
style issues at https://test.civicrm.org/job/CiviCRM-Style/1039/Civilint/

eileenmcnaughton commented:
So to the question of unit tests I think there needs to be some form of test that if a token exists it can be listed & rendered. We have similar tests in the `TokenConsistencyTest` class - eg https://github.com/civicrm/civicrm-core/blob/8d8b739113841b8121f6e22020e2b5c0efe472ab/tests/phpunit/CRM/Utils/TokenConsistencyTest.php#L872 (ignore the first couple of lines in that test)

There is also some functional code in https://github.com/civicrm/civicrm-core/pull/30451/files#diff-3aa6439386b5cc7a356fce232e9fd1986caf5a670d8869735cf4e0b9c0aed8feR24 which would be good to test (population on created_id) - the hardest question with these things is always 'where do I put this' - we talked about this just now & are leaning towards a new place - see https://chat.civicrm.org/civicrm/pl/rxuaa5afuin58mfokfiiaz8w3a - but if that gets sucked into mole-made-bike-shed-worm-hole-with-no-tale-or-tail then `CRM_Core_BAO_SiteTokenTest` would be a sane place to put it 

Lastly it would be good to test at least one message template, as ships, with a site token - I would probably piggy back on an already existing test like this one https://github.com/civicrm/civicrm-core/blob/789576b27d4f0880cf8ed1bd45e5b5adb0bcccdb/tests/phpunit/CRM/Contribute/Form/ContributionTest.php#L538 & just create/ update the site-token before it runs & then check it in the test. You would need to revert the siteToken in the `tearDown()`

eileenmcnaughton commented on CRM/Core/SiteTokens.php:
```diff
@@ -0,0 +1,105 @@
+  /**
+   * Get the site tokens available for the domain.
+   *
+   * This function will be made protected soon... FIXME: this line is copied from CRM_Core_DomainTokens::getDomainTokenValues, so not sure if it's true here. -TwoMice
```
```suggestion

```

eileenmcnaughton commented on CRM/Core/SiteTokens.php:
```diff
@@ -0,0 +1,105 @@
+   *
+   * @todo - make this non-static & protected. Remove last deprecated fn that calls it.
+   */
+  public static function getSiteTokenValues(?int $domainID = NULL, bool $html = TRUE): array {
```
```suggestion
  private function getSiteTokenValues(?int $domainID = NULL, bool $html = TRUE): array {
```

this was only ever public / static on domain tokens as part of the transition  from `CRM_Core_Token` - make everything private that does not need to be public or else when we want to alter them someone will have some weird & wonderful code calling it directly & will think it's our fault if it breaks

eileenmcnaughton commented:
@twomice one conversation we had was that once we put this out there someone will want to make it recursive - I did a bit of a braindump on thoughts about that & approaches on the gitlab https://lab.civicrm.org/dev/core/-/issues/3682#note_166290 - I think how that relates to this PR is that it would be good to add a comment in an appropriate place in the site-tokens code redirectly anyone who might be thinking about adding recursivity back to the gitlab discussion so at least they are informed by whatever the discussion evolves as at that time

eileenmcnaughton commented on CRM/Core/SiteTokens.php:
```diff
@@ -0,0 +1,104 @@
+  private function getSiteTokenValues(?int $domainID = NULL, bool $html = TRUE): array {
+
+    $ret = [];
+    $siteTokens = \Civi\Api4\SiteToken::get(TRUE)
```
I was thinking about this & I think the siteTokens should be in `\Civi::$statics` as this could be done quite a few times in an iteration.

I don't think we want a heavier cache cos if we got to the point of recursive then `{domain.now}` would get stale in an hour-long cache

eileenmcnaughton commented on Civi/Core/Container.php:
```diff
@@ -389,6 +389,10 @@ public function createContainer() {
       'CRM_Core_DomainTokens',
       []
     ))->addTag('kernel.event_subscriber')->setPublic(TRUE);
+    $container->setDefinition('crm_site_tokens', new Definition(
+      'CRM_Core_SiteTokens',
+      []
+    ))->addTag('kernel.event_subscriber')->setPublic(TRUE);
```
```suggestion
  
```
the other one got merged

eileenmcnaughton commented:
@twomice FYi - I'm gonna put up a commit with the schema & upgrade stuff separately - I have some changes & also good to get that part merged as it goes stale / is hard to get past tests

twomice commented:
Okay, thanks @eileenmcnaughton . I'm moving on this as quickly as I can, so if you're able to get action more quickly that way, that's a good thing.

eileenmcnaughton commented:
@twomice https://github.com/civicrm/civicrm-core/pull/30518 - let's see if it passes

eileenmcnaughton commented:
@twomice I just rebased this over other patches - from my point of view if this still runs OK there is nothing blocking it being merged & we can tweak further post-merge

Ideally we would follow up with
- a pre-upgrade message pointing to the docs you have been writing
- tests
- some caching in that eval function (just `\Civi::$statics` level)

twomice commented:
@eileenmcnaughton Okay, I've made a few small changes to the language ("Token Name" is now "Key") and added help text in the Site Token edit/create form.

## Note on validation
I'd like to see more validation in the submission of the form, and @colemanw said he has ideas about getting php-schema-level validation and letting those messages be seen by the user of a Form Builder form. Specifically, I can already see it would be nice to enforce:

- `name` should be limited to valid `{token.[name]}` values (i.e. alphanumeric without spaces or punctuation)  
  Right now, the form will let you create a token like `{site.foo bar}`, which will of course produce unpredictable results (and perhaps errors) in actual usage.
- `name` should be uniqe per domain.  
  This is currently enforced at the DB level, and the DB won't allow it, but there's no user-visible message; the token is simply not saved.
- `label` should be unique per domain.  
  It would be nonsensical to have two Site Tokens with the same label "Foo" and "Foo". This is currently not enforced anywhere; the user can do this, and the duplicate-labeled token is saved and made available for use.

## PR status
In any case, I think this is ready for your review and hopefully meets your approval for merging. I've removed "WIP" from the PR title.

eileenmcnaughton commented on ang/afform/afsearchSiteTokens.aff.html:
```diff
@@ -0,0 +1,9 @@
+<div class="af-markup">
+  <div class="help">
+  <div>{{:: ts('Site Tokens allow you to define pre-set snippets of content which you can then insert via CiviCRM\'s standard token-handling features. Site Tokens marked as "reserved" are also inserted automatically in specific places within CiviCRM features, such as Message Templates.')}} <a class="crm-doc-link no-popup" href="https://docs.civicrm.org/user/en/latest/common-workflows/tokens-and-m ...
```
I need to check the right way to add a link - this has the language hard-coded - maybe @colemanw  or @mlutfy knows

eileenmcnaughton commented on ang/afform/afsearchSiteTokens.aff.json:
```diff
@@ -0,0 +1,31 @@
+    "icon": "",
+    "server_route": "civicrm/admin/sitetoken",
+    "permission": [
+        "edit system workflow message templates"
```
This permission makes sense - I'm guessing the search display would wind up with 'administer CiviCRM' by default - which might be OK or might be too restrictive? I think we should probably also define this permission in CRM_Core_Permission as the right permission for create / edit these in general (ie it would also apply to the api)

eileenmcnaughton commented on ang/afform/afsearchSiteTokens.aff.json:
```diff
@@ -0,0 +1,31 @@
+        "label": "Site Tokens",
+        "weight": 10
+    },
+    "modified_date": "2024-06-16 15:50:36",
```
seems a bit weird to include modified date here - but I would defer to @colemanw on that

eileenmcnaughton commented:
I'm going to push any follow ups I identify to the check list I wrote here https://lab.civicrm.org/dev/core/-/issues/3682 rather than require them to be all solved pre-merge

eileenmcnaughton commented:
In my testing on the demo site this worked with some gaps

1) the message templates did not include the new token
2) the reserved site token was not created

I am assuming that some sql regeration will fix these & my overall position at this point is that this should be merged & some things done as follow ups (which I'm trying to track on the gitlab now) . I'm going to merge this & then try to do the regeneration as a follow up PR

eileenmcnaughton commented:
@twomice - I've done the regen PR https://github.com/civicrm/civicrm-core/pull/30529 - but I'm just trying to see where the message_header token stub is created - could that have gotten lost along the way?

eileenmcnaughton commented:
This adds a unique index on the label+domain_id per comments above https://github.com/civicrm/civicrm-core/pull/30530
